### PR TITLE
AP_GPS: Add NMEA THS message.

### PR DESCRIPTION
**Pull request by phamelin**
Add support for NMEA THS message, which is basically the same as HDT but with a data qualifier.

### CODE REVIEW

**Comment by peterbarker** [comment]
Doesn't this kind of break things for people running devices not emitting this message?

**Comment by khancyr** [comment]
@peterbarker, it doesn't seems to break the previous behavior, just add the parsing of another message.

@phamelin could you split your commit in two : one for AP_GPS, the other for AP_HAL_SITL ?

I am not knowledgable enough on NMEA but I don't think we should change the SITL part.

**Comment by phamelin** [comment]
@khancyr I was also concerned with the SITL part. Currently, the HDT message is sent even if there's no gps lock. That's why I replaced HDT with THS in SITL because this one has a data qualifier. It would be nice if we could generate either HDT or THS in SITL, but this would imply additionnal SIM_* parameters. What we could do is replace SIM_GPS_HDG with two parameters: SIM_GPS_HDT and SIM_GPS_THS. I don't know if we want to go that way. Anyway, I will split the commit and then we will be able to discuss the modifications.

**Comment by phamelin** [comment]
The pull request is now splitted in two commits.

**Comment by yanivasy** [comment]
I think the time/week bug in NMEA parsing should be fixed as well
https://github.com/ArduPilot/ardupilot/issues/14298

**Comment by phamelin** [comment]
@WickedShell Other than resolving the conflict with the master branch, is there anything I can do to make this pull request go ahead?

**Comment by khancyr** [comment]
@phamelin could you rebase it ? I will bring it on devcall tomorrow. Hopefully, the NMEA part could be merge easily, and we could discuss the SITL part.

**Comment by phamelin** [comment]
@khancyr I apologize for the late reply. I have been very busy recently. I just did the rebase. Thank you for handling this.

File: libraries/AP_GPS/AP_GPS_NMEA.h [hunk 1]
```diff
@@ -146,7 +147,7 @@ class AP_GPS_NMEA : public AP_GPS_Backend
     uint32_t _last_RMC_ms;
     uint32_t _last_GGA_ms;
     uint32_t _last_VTG_ms;
-    uint32_t _last_HDT_ms;
+    uint32_t _last_HDT_THS_ms{0};
```
Comment (tridge) [inline libraries/AP_GPS/AP_GPS_NMEA.h:150]: we don't need to init this to zero as we auto-init all allocated objects to zero. Adding the init consumes flash space unnecessarily

File: libraries/AP_HAL_SITL/sitl_gps.cpp [hunk 2]
```diff
@@ -756,7 +756,7 @@ void SITL_State::_update_gps_nmea(const struct gps_data *d, uint8_t instance)
                      dstring);
 
     if (_sitl->gps_hdg_enabled[instance]) {
-        _gps_nmea_printf(instance, "$GPHDT,%.2f,T", d->yaw);
+        _gps_nmea_printf(instance, "$GPTHS,%.2f,%c,T", d->yaw, d->have_lock ? 'A' : 'V');
```
Comment (tridge) [inline libraries/AP_HAL_SITL/sitl_gps.cpp:759]: let's make it so that SIM_GPS_HDG==1 means HDT and SIM_GPS_HDG==2 means THS, so we can test both

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
also, can you add a comment in the code explaining which GPS vendor includes the THS message?

**Comment by phamelin** [comment]
@tridge I just pushed the requested changes. Thank you.

**Review by khancyr — APPROVED** [review]
seems good

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
nice PR, thanks!

**Pull request merged**